[SPARK-12063][SQL] Use number in group by clause to refer to columns

dereksabryfb opened the pull request:
If there is a number n in a group by clause, the nth column in the select clause is used to modify the group by clause to refer to this column instead of the number.

eg.
select a,b from c group by 1,2
becomes
select a,b from c group by a,b

Commit:
Use number to refer to columns in group by clause

AmplabJenkins commented:
Can one of the admins verify this patch?

marmbrus commented:
Thanks for working on this.  This seems resonable to support, but I have two suggestions:
 - I would probably implement this as a rule in the [`Analyzer`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala) so that it is not specific to the Hive parser.
 - Please add a unit test, probably in `SQLQuerySuite`

rxin commented:
@dereksabryfb   you should also add the email you used in your git commit to your github profile so it shows up on github.

Commit:
Group by Column Number for Spark SQL

dereksabryfb commented:
Thanks for your feedback! The change is now implemented as a rule in the Analyzer and a unit test has been added. Let me know if there are any other changes I need to make.

Commit:
minor style change

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -178,6 +179,27 @@ class Analyzer(
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        Aggregate(groups.map(group => group match {
+          case g if g.prettyString forall Character.isDigit =>
```
I'm not sure we want to match on the pretty string as I think this would also trigger for things like `"1"`, instead I'd consider matching on a Literal of NumericType.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -178,6 +179,27 @@ class Analyzer(
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        Aggregate(groups.map(group => group match {
```
Nit: would you mind assigning this to an intermediate variable.  this is a little hard to parse visually.

marmbrus commented:
Are there other cases that we should handle here?  Can you do the same thing in ORDER BY normally?

dereksabryfb commented:
Thanks for the feedback! I'm making the changes you suggested. With respect to the 'ORDER BY' clause, this looks to be semi-handled by the ResolveSortReferences in the Analyzer; because in standard SQL literals are allowed in the order by clause, this rule just sorts by the literal value '1'; In HiveQL, '1' is interpreted as a column in the same way it is in the group by clause.

I can add the case for a Sort() with an IntegerType. I assume there is no one currently relying on how 'sort by 1' currently functions.

marmbrus commented:
We'll have to note the change in the release notes, but since its a no-op to sort by a constant I think we can safely change behavior here.

Commit:
Add Sort() case

dereksabryfb commented:
Added a case for sort

marmbrus commented:
ok to test

SparkQA commented:
**[Test build #47534 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47534/consoleFull)** for PR 10052 at commit [`8a5a4f6`](https://github.com/apache/spark/commit/8a5a4f63fc66792e924f4f3355df357815aae13b).

SparkQA commented:
**[Test build #47534 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47534/consoleFull)** for PR 10052 at commit [`8a5a4f6`](https://github.com/apache/spark/commit/8a5a4f63fc66792e924f4f3355df357815aae13b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47534/
Test FAILed.

dereksabryfb committed:
scala style

dereksabryfb commented:
Apologies, I haven't been able to run ./dev/run-tests, getting the following exception: http://pastebin.com/L0p0sjtJ

so I wasn't able to pick up the style issues, and I'm not sure if there's more that the build doesn't flag.

marmbrus commented:
I'd try the following locally `build/sbt scalastyle test:scalastyle catalyst/test sql/test`.

Each of those commands can be run separately too and you can use ~ to rerun whenever something changes to iterate more quickly `build/sbt ~scalastyle`

SparkQA commented:
**[Test build #47538 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47538/consoleFull)** for PR 10052 at commit [`bd453d5`](https://github.com/apache/spark/commit/bd453d5f6744aa8fdd03b5ee2ecd44b471165eb4).

SparkQA commented:
**[Test build #47538 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47538/consoleFull)** for PR 10052 at commit [`bd453d5`](https://github.com/apache/spark/commit/bd453d5f6744aa8fdd03b5ee2ecd44b471165eb4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47538/
Test FAILed.

dereksabryfb commented:
Looks like it fails on the query "SELECT a, count(2) FROM testData2 GROUP BY a, 2" because 2 now refers to the column count(2) which is not valid syntax for a group by clause in hive. How should I go about resolving this? The purpose of the test is to see that literals in the group by clause don't modify the results, but the purpose of this patch is to do the opposite of that. I could modify the offending case, but I feel the whole test may be irrelevant with this patch.

marmbrus commented:
We should probably throw an `AnalysisException` for this if they use a column ordinal that refers to an aggregate expression.  The fact that we make it all the way to the execution is pretty confusing to a user.

Regarding the test, we can probably remove it.

dereksabryfb committed:
Remove literal in group by clause test, as this functioanlity has changed

SparkQA commented:
**[Test build #47565 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47565/consoleFull)** for PR 10052 at commit [`e4edc31`](https://github.com/apache/spark/commit/e4edc31426788f167be81bdeb72aea64227c17bf).

dereksabryfb committed:
aggregate in group by clause causes failure in execution

dereksabryfb commented:
I removed the offending test.

I found that there was no `AnalysisException` thrown even if there was an explicit aggregate in the group by clause (e.g. select a from b group by count(a)), and it would fail in the same way, so I added the check to `CheckAnalysis` ; if you think this is out of the scope of this pull request (since it isn't strictly to do with a number reference), I can create a new task and attach just that commit to it.

Thanks again for your feedback.

SparkQA commented:
**[Test build #47566 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47566/consoleFull)** for PR 10052 at commit [`09b3f77`](https://github.com/apache/spark/commit/09b3f77630be93c87abc6b64c4754b7cb32495e7).

SparkQA commented:
**[Test build #47566 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47566/consoleFull)** for PR 10052 at commit [`09b3f77`](https://github.com/apache/spark/commit/09b3f77630be93c87abc6b64c4754b7cb32495e7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47566/
Test FAILed.

SparkQA commented:
**[Test build #47565 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47565/consoleFull)** for PR 10052 at commit [`e4edc31`](https://github.com/apache/spark/commit/e4edc31426788f167be81bdeb72aea64227c17bf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47565/
Test FAILed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -178,6 +179,46 @@ class Analyzer(
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        val newGroups = groups.map {
+          case group if group.isInstanceOf[Literal] && group.dataType.isInstanceOf[IntegralType] =>
```
```case Literal(index: Int) => ``` is easier. It also eliminates the need for ```group.toString.toInt```

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -178,6 +179,46 @@ class Analyzer(
+        Aggregate(newGroups, aggs, child)
+      case Sort(ordering, global, child) =>
+        val newOrdering = ordering.map {
+          case o if o.child.isInstanceOf[Literal] && o.dataType.isInstanceOf[IntegralType] =>
```
Same comment as above.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -178,6 +179,46 @@ class Analyzer(
+      case Sort(ordering, global, child) =>
+        val newOrdering = ordering.map {
+          case o if o.child.isInstanceOf[Literal] && o.dataType.isInstanceOf[IntegralType] =>
+            val newExpr = child.asInstanceOf[Project].projectList(o.child.toString.toInt - 1)
```
Tests are currently failing to due this cast. A ```Sort```'s child doesn't have to be a ```Project```, use the LogicalPlan ```output``` attributes to get the proper attribute.

hvanhovell commented:
@dereksabryfb are you still working on this?

srowen commented:
Ping @dereksabryfb -- are you working on this? or else close it

hvanhovell commented:
I think this one can be closed, it has been implemented in https://github.com/apache/spark/pull/11846

JoshRosen commented:
Yep, it looks like this PR has been subsumed by #11846. @dereksabryfb, could you please close this pull request? Thanks!
